internal/workflow/activity/update_watermark.go

```diff
@@ -63,13 +65,30 @@ func (a *UpdateWatermark) execute(ctx context.Context, request *UpdateWatermarkR
 	tag := a.config.GetEffectiveBlockTag(request.Tag)
 	logger.Info("Updating watermark",
 		zap.Uint32("tag", tag),
+		zap.Uint64("validate_since", request.ValidateSince),
 		zap.Uint64("height", request.BlockHeight))
 
-	block, err := a.metaStorage.GetBlockByHeight(ctx, tag, request.BlockHeight)
+	validateStart := request.BlockHeight - 1
+	if request.ValidateSince > 0 {
+		if request.ValidateSince >= request.BlockHeight {
+			return nil, xerrors.Errorf("ValidateSince %d should be smaller than BlockHeight %d",
+				request.ValidateSince, request.BlockHeight)
+		}
+		validateStart = request.ValidateSince
+	}
+	if validateStart <= 0 {
+		validateStart = 1
+	}
+	blocks, err := a.metaStorage.GetBlocksByHeightRange(ctx, tag, validateStart, request.BlockHeight+1)
```

Comment:
The GetBlocksByHeightRange will sort the blocks, so trying to get the blocks in reverse order is not allowed, so it'll fail here.